### PR TITLE
Add tracing

### PR DESCRIPTION
**Pull request by estrada9166**
The idea is to have a tracer on the fields of the operation, before landing it, I have few questions...

### Questions:

+ Shall we have startTime and endTime? do you think those are in the right place?
+ How we should display the tracing result?

### Missing: 

- [x] UT
- [x] Update documentation

### Result
```
{
  "startTime": 42511540659488,
  "endTime": 42515156653108,
  "duration": 3615993620,
  "tracingResolvers": [
    {
      "path": ["person"],
      "parentType": "Query",
      "returnType": "Person!",
      "startTime": 42515155683053,
      "endTime": 42515155770657,
      "duration": 87604
    },
    {
      "path": ["person", "firstname"],
      "parentType": "Person",
      "returnType": "String!",
      "startTime": 42515156403990,
      "endTime": 42515156411814,
      "duration": 7824
    },
    {
      "path": ["person", "lastname"],
      "parentType": "Person",
      "returnType": "String!",
      "startTime": 42515156578296,
      "endTime": 42515156580820,
      "duration": 2524
    }
  ]
}
```

I think for the moment that is it!

Thanks

### CODE REVIEW

**Comment by geek** [comment]
@estrada9166 thanks for the PR.

Is the intention to provide timing information around the execution of the handler along with the graphql resolver... If so then I'd move the startTime as a local variable in the handler and create it somewhere near the top of the handler, like https://github.com/geek/graphi/pull/22/files#diff-6d186b954a58d5bb740f73d84fe39073R133 

We can remove the end time from the results if you want and just keep duration and start time, up to you on that.

**Comment by estrada9166** [comment]
This'll be the new result:
```json
{
  "duration": 18349299,
  "tracingResolvers": [
    {
      "fieldName": "person",
      "path": ["person"],
      "parentType": "Query",
      "returnType": "Person!",
      "duration": 321198
    },
    {
      "fieldName": "firstname",
      "path": ["person","firstname"],
      "parentType": "Person",
      "returnType": "String!",
      "duration": 11605
    },
    {
      "fieldName": "lastname",
      "path": ["person","lastname"],
      "parentType": "Person",
      "returnType": "String!",
      "duration": 4944
    }
  ]
}
```

also, what do you think about displaying the result like this using [cli-table](https://www.npmjs.com/package/cli-table)
screen recording 2019-02-19 at 02 15 pm

**Comment by estrada9166** [comment]
@geek almost everything is done! I'm missing the UT... any hint how should I do it?

Thanks!

File: package.json [hunk 1]
```diff
@@ -37,9 +37,10 @@
     "wreck": "14.x.x"
   },
   "dependencies": {
+    "apollo-server-module-graphiql": "1.4.x",
     "boom": "7.x.x",
+    "cli-table": "^0.3.1",
```
Comment (geek) [inline package.json]: can you make this `0.3.x`, for consistency

**Comment by geek** [comment]
I really like the table output, any thoughts on sorting it by least duration to largest, that way the trouble makers are all together?

**Comment by estrada9166** [comment]
I really like it sorted! just pushed those changes! 

Just missing the UT, any idea how should I do it?

**Comment by geek** [comment]
@estrada9166 what does UT stand for :) user time?

**Comment by lloydbenson** [comment]
I'm guessing Unit Test.

**Comment by estrada9166** [comment]
Sorry! 

Yes! it's Unit Test.

**Comment by geek** [comment]
@estrada9166 I added a basic test, see what you think of this direction

File: lib/index.js [hunk 2]
```diff
@@ -208,10 +208,7 @@ internals.graphqlHandler = async function (request, h) {
         ]);
       });
 
-    if (tracingResult.tracingResolvers.length) {
```
Comment (estrada9166) [inline lib/index.js]: The only thing that I'm afraid of removing this validation, is that while using `GraphiQL` when you load it the first time or refresh it, there is a request so, it'll display on console an empty table.

Comment (estrada9166) [inline lib/index.js]: The only thing that I'm afraid of removing this validation, is that while using `GraphiQL` when you load it the first time or refresh it, there is a request so, it'll display on console an empty table.

Comment (geek) [reply lib/index.js]: That sounds reasonable, can you add a test case to cover that condition? I added one that I think tests what you are describing, but the tracingResolvers has a length.

**Comment by geek** [comment]
@estrada9166 is this ready for final review?!

**Comment by estrada9166** [comment]
@geek Yes! it's ready now :)

**Comment by geek** [comment]
@estrada9166 sorry for the delay. I updated graphi with event support for the pre/post field resolvers. This then allowed me to take your tracing work and make it into a hapi plugin. I invited you to the repo, feel free to take it over and move it under your name if you'd like, publish, and do whatever you want with. 

The main reason for doing this is that it will allow your tracing support to work and also allow you full control over its features. I figure this will be nicer than having to publish new graphi versions when we need tracing changes.

**Comment by estrada9166** [comment]
>  I updated graphi with event support for the pre/post field resolvers.

Nice! agree with this!

> I invited you to the repo, feel free to take it over and move it under your name if you'd like, publish, and do whatever you want with.

No problem! we can leave it like it is :)

Thank you so much!

**Pull request closed without merging**